File: 0047-permutations-ii/0047-permutations-ii.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        def permute(route):
            if len(route) == len(nums):
                result.append(list(route))
                return
            
            for i in range(len(nums)):
                if chk[i] or (i > 0 and nums[i] == nums[i - 1] and not chk[i - 1]):
                    continue
                chk[i] = True
                route.append(nums[i])
                permute(route)
                route.pop()
                chk[i] = False


        chk = [False for i in range(len(nums))]
        result = list()
        nums.sort()

        permute(list())
        return result

File: 0047-permutations-ii/0047-permutations-ii.java
class Solution {
    List<List<Integer>> result = new ArrayList<>();
    Set<List<Integer>> record = new HashSet<>();
    Set<Integer> num = new HashSet<>();

    private void backtrack(int[] nums, List<Integer> route) {
        if (route.size() == nums.length && !record.contains(route)) {
            List<Integer> temp = new ArrayList<Integer>(route);
            record.add(temp);
            result.add(temp);
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (num.contains(i)) {
                continue;
            }
            route.add(nums[i]);
            num.add(i);
            backtrack(nums, route);
            num.remove(i);
            route.remove(route.size() - 1);
        }
    }

    public List<List<Integer>> permuteUnique(int[] nums) {
        Arrays.sort(nums);
        backtrack(nums, new ArrayList<>());
        return result;
    }
}